src/types/osstr.rs

```diff
@@ -0,0 +1,191 @@
+use crate::types::PyString;
+#[cfg(not(windows))]
+use crate::PyNativeType;
+use crate::{
+    ffi, AsPyPointer, FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python,
+    ToPyObject,
+};
+use std::borrow::Cow;
+use std::ffi::{OsStr, OsString};
+#[cfg(not(windows))]
+use std::os::raw::c_char;
+
+impl ToPyObject for OsStr {
+    fn to_object(&self, py: Python) -> PyObject {
+        // If the string is UTF-8, take the quick and easy shortcut
+        if let Some(valid_utf8_path) = self.to_str() {
+            return PyString::new(py, valid_utf8_path).into();
+        }
+
+        // Take advantage of Python's surrogateescape encoding to support arbitrary byte
+        // sequences instead of limiting ourselves to UTF-8
+
+        // All targets besides windows support the std::os::unix::ffi::OsStrExt API:
+        // https://doc.rust-lang.org/src/std/sys_common/mod.rs.html#59
+        #[cfg(not(windows))]
+        {
+            let bytes = std::os::unix::ffi::OsStrExt::as_bytes(self);
+
+            let ptr = bytes.as_ptr() as *const c_char;
+            let len = bytes.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is this the appropriate function? Does it use the surrogateescape error
+                // handler? Should it be PyUnicode_DecodeLocaleAndSize instead maybe?
+                let pystring = ffi::PyUnicode_DecodeFSDefaultAndSize(ptr, len);
+                py.from_owned_ptr::<PyString>(pystring).into()
+            }
+        }
+
+        #[cfg(windows)]
+        {
+            let wstr: Vec<u16> = std::os::windows::ffi::OsStrExt::encode_wide(self).collect();
+
+            let ptr = wstr.as_ptr();
+            let len = wstr.len() as ffi::Py_ssize_t;
```

Comment:
BTW, you can just do `ffi::PyUnicode_FromWideChar(ptr, wstr.len() as _)`

Reply:
I did it for consistency with the block above. Also, with rustfmt, that change would actually increase the number of lines. Hence I'm gonna keep it like before